### PR TITLE
Added Cluster ID to sendgrid email template

### PR DESCRIPTION
**Pull request by akshayashu**
The new email template will look like this:

sendgrid

Note: I used a new library github.com/openshift/api/config/v1 to be able to use clusterVersions

Signed-off-by: akkhanna [akkhanna@redhat.com](mailto:akkhanna@redhat.com)

### CODE REVIEW

**Comment by nb-ohad** [comment]
@akshayashu @bindrad I think the current template is unnecessarily complicated and results in HTML that does not reflect the correct structure of the email. One example is the fact that each line of the list is its own unordered list (UL) containing a single list item (LI). Also, the cluster id should move into the alert description, not the instance data list.

I would suggest simplifying the template using the following suggestion:
```
<<!DOCTYPE html>
<html>
<head>
    <title>Alerts</title>
</head>
<body itemscope itemtype="http://schema.org/EmailMessage">
    <p>
        Hello!
        <br><br>
        This notification is for your OpenShift managed cluster running OpenShift Data Foundation.
        <br><br>
        <strong>
        {{ $firstAlert := (index .Alerts.Firing 0).Labels }}

        {{ if eq $firstAlert.alertname "CephClusterCriticallyFull" }}
            Ceph Cluster is Critically Full
            <br><br>
            Your storage cluster ({{ $firstAlert.clusterId }})  utilization has crossed 80% and will move into a read-only state at 85%! Please free up some space or if possible expand the storage cluster immediately to prevent any service access issues.
        {{ end }}
        {{ if eq $firstAlert.alertname "CephOSDNearFull" }}
            Ceph OSD is Nearly Full
            <br><br>
            The utilization of an OSD in your cluster ({{ $firstAlert.clusterId }}) has exceeded 75%. Please free up some space or expand the storage cluster if possible. If an OSD reaches 85% utilization it will enter a read-only state and service interruptions are possible.
        {{ end }}
        {{ if eq $firstAlert.alertname "CephOSDCriticallyFull" }}
            Ceph OSD is Critically Full
            <br><br>
            The utilization of an OSD in your cluster ({{ $firstAlert.clusterId }}) has exceeded 80%. Please free up some space immediately or expand the storage cluster if possible. Failure to address this issue will lead to service interruptions.
        {{ end }}
        {{ if eq $firstAlert.alertname "PersistentVolumeUsageNearFull" }}
            Persistent Volume Usage is Nearly Full
            <br><br>
            The utilization of one or more of the PVs in your cluster ({{ $firstAlert.clusterId }}) has exceeded 75%. Please free up some space or expand the PV if possible. Failure to address this issue may lead to service interruptions.
        {{ end }}
        {{ if eq $firstAlert.alertname "PersistentVolumeUsageCritical" }}
            Persistent Volume Usage Critical
            <br><br>
            The utilization of one or more of the PVs in your cluster ({{ $firstAlert.clusterId }}) has exceeded 85%. Please free up some space immediately or expand the PV if possible. Failure to address this issue may lead to service interruptions.
        {{ end }}
        {{ if eq $firstAlert.alertname "CephClusterReadOnly" }}
            Ceph Cluster is in Read Only state
            <br><br>
            Your storage cluster ({{ $firstAlert.clusterId }}) utilization has crossed 85%, as a result, access is now in a read-only state! Please free up some space or if possible expand the storage cluster immediately to prevent any additional service access issues.
        {{ end }}
        {{ if eq $firstAlert.alertname "CephClusterNearFull" }}
            Ceph Cluster is Nearly Full
            <br><br>
            Your storage cluster ({{ $firstAlert.clusterId }}) utilization has crossed 75% and will move into a read-only state at 85%. Please free up some space or if possible expand the storage cluster immediately to prevent any service access issues.
        {{ end }}
        <br>

        {{ range .Alerts.Firing }}
            <ul>    
                <li>PVC Name: {{ .Labels.persistentvolumeclaim }}</li>
                <li>Namespace: {{ .Labels.namespace }}</li>
            </ul>
	    <br>
        {{ end }}
        
        </strong>
        <br>
        
        If you have any questions, please <a clicktracking=off href="https://access.redhat.com/support/contact/technicalSupport/">contact us</a>. Review the <a clicktracking=off href="https://access.redhat.com/support/policy/support_process">support process</a> for guidance on working with Red Hat support.
        <br><br>
        Thank you for choosing Red Hat OpenShift Data Foundation,
        <br>
        ODF SRE
    </p>
</body>
</html>
```

**Comment by akshayashu** [comment]
@nb-ohad We will have issue with the template you suggested, there are alerts that don't contain PVC name, it will be blank in that case.

**Comment by nb-ohad** [comment]
@akshayashu I think `with` will work better then an `if`

```
{{ with .Labels.persistentvolumeclaim }}<li>PVC Name: {{ . }}</li>{{ end }}
```

File: controllers/managedocs_controller.go [hunk 1]
```diff
@@ -1042,6 +1044,14 @@ func (r *ManagedOCSReconciler) reconcilePrometheus() error {
 			}
 		}
 
+		// getting clusterID
+		clusterVersion := &configv1.ClusterVersion{}
+		if err := r.UnrestrictedClient.Get(context.Background(), types.NamespacedName{Name: "version"}, clusterVersion); err != nil {
+			return err
+		}
```
Comment (nb-ohad) [inline controllers/managedocs_controller.go:1051]: Why do we need the `UnrestrictedClient`? and even if we need it we have a `r.UnrestrictedClient` that you can use

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/managedocs_controller.go (hunk 1 above)
Comment (ezio-auditore) [reply controllers/managedocs_controller.go:1051]: @nb-ohad I think you meant r.unrestrictedGet, right?

Comment (nb-ohad) [reply controllers/managedocs_controller.go:1051]: Correct, updating the comment

File: templates/customernotification.html [hunk 2]
```diff
@@ -10,73 +10,46 @@
...
+                <br><br>
+                Your storage cluster ({{ .clusterId }})  utilization has crossed 80% and will move into a read-only state at 85%! Please free up some space or if possible expand the storage cluster immediately to prevent any service access issues.
+            {{ else if eq .alertname "CephOSDNearFull" }}
+                Ceph OSD is Nearly Full
+                <br><br>
+                The utilization of an OSD in your cluster ({{ .clusterId }}) has exceeded 75%. Please free up some space or expand the storage cluster if possible. If an OSD reaches 85% utilization it will enter a read-only state and service interruptions are possible.
+            {{ else if eq .alertname "CephOSDCriticallyFull" }}
+                Ceph OSD is Critically Full
+                <br><br>
+                The utilization of an OSD in your cluster ({{ .clusterId }}) has exceeded 80%. Please free up some space immediately or expand the storage cluster if possible. Failure to address this issue will lead to service interruptions.
```
[the remark this answers is not in the document]
Comment (nb-ohad) [reply templates/customernotification.html:25]: Only If we are not configuring these alerts to sendgrid

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: templates/customernotification.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (akshayashu) [reply templates/customernotification.html:25]: We are not configuring these alerts

[the remark this answers is not in the document]
Comment (nb-ohad) [reply templates/customernotification.html:25]: so remove

File: templates/customernotification.html [hunk 3]
```diff
@@ -10,73 +10,46 @@
...
-                        <li>PVC Name: {{ .Value }}</li>
-                    {{ end }}
-                    {{if eq .Name "namespace" }}
-                        <li>Namespace: {{ .Value }}</li>
-                    {{ end }}
-                </ul>
-            {{ end }}
+            <ul>    
+                {{ with .Labels.persistentvolumeclaim }}<li>PVC Name: {{ . }}</li>{{ end }}
+                {{ with .Labels.persistentvolumeclaim }}<li>Namespace: {{ .Labels.namespace }}</li>{{ end }}
```
Comment (nb-ohad) [inline templates/customernotification.html:41]: I this context `.` will be `.Labels.persistentvolumeclaim` so `.Labels.namespace` will become `.Labels.persistentvolumeclaim.Labels.namespace` which I believe is not what you intendant 

The solution will be to use an `if` the encompass the entire UL based on the value of `.Labels.persistentvolumeclaim`

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/managedocs_controller.go [hunk 4]
```diff
@@ -1043,14 +1045,22 @@ func (r *ManagedOCSReconciler) reconcilePrometheus() error {
...
+
+        r.prometheus.Spec = desired.Spec
+        r.prometheus.Spec.ExternalLabels["clusterId"] = string(clusterVersion.Spec.ClusterID)
+        r.prometheus.Spec.Alerting.Alertmanagers[0].Namespace = r.namespace
+        r.prometheus.Spec.AdditionalAlertRelabelConfigs = &corev1.SecretKeySelector{
+            LocalObjectReference: corev1.LocalObjectReference{
+                Name: alertRelabelConfigSecretName,
+            },
+            Key: alertRelabelConfigSecretKey,
+        }
```
Comment (nb-ohad) [inline controllers/managedocs_controller.go:1063]: Something is wrong with the indentation here

**Review by nb-ohad — CHANGES_REQUESTED** [review]

**Comment by nb-ohad** [comment]
@akshayashu in addition to the PR review, please squash into 2 commits, 1 for vendor changes and another one for the code changes

File: controllers/managedocs_controller.go (hunk 4 above)
Comment (akshayashu) [reply controllers/managedocs_controller.go]: corrected

**Comment by akshayashu** [comment]
Changed to 2 commits (one for vendor and one for code)

**Comment by nb-ohad** [comment]
@akshayashu You have conflicts, please rebase

**Comment by akshayashu** [comment]
Rebased

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/managedocs_controller.go [hunk 5]
```diff
@@ -1040,7 +1042,14 @@ func (r *ManagedOCSReconciler) reconcilePrometheus() error {
 			}
 		}
 
+		clusterVersion := &configv1.ClusterVersion{}
+		clusterVersion.Name = "version"
+		if err := r.unrestrictedGet(clusterVersion); err != nil {
```
[the remark this answers is not in the document]
Comment (akshayashu) [reply controllers/managedocs_controller.go:1047]: We see an error while using `r.get` 
Failed to watch *v1.ClusterVersion: failed to list *v1.ClusterVersion: clusterversions.config.openshift.io is forbidden: User "system:serviceaccount:openshift-storage:ocs-osd-deployer" cannot list resource "clusterversions" in API group "config.openshift.io" at the cluster scope

[the remark this answers is not in the document]
Comment (nb-ohad) [reply controllers/managedocs_controller.go:1047]: Seems like you are missing some RBAC

File: controllers/managedocs_controller.go [hunk 6]
```diff
@@ -172,6 +173,7 @@ type ManagedOCSReconciler struct {
 // +kubebuilder:rbac:groups="network.openshift.io",namespace=system,resources=egressnetworkpolicies,verbs=create;get;list;watch;update
 // +kubebuilder:rbac:groups="coordination.k8s.io",namespace=system,resources=leases,verbs=create;get;list;watch;update
 // +kubebuilder:rbac:groups="odf.openshift.io",namespace=system,resources=storagesystems,verbs=list;watch;delete
+// +kubebuilder:rbac:groups="config.openshift.io",resources=clusterversions,verbs=get
```
Comment (bindrad) [inline controllers/managedocs_controller.go:176]:
```suggestion
// +kubebuilder:rbac:groups="config.openshift.io",resources=clusterversions,verbs=get;watch
```

Comment (akshayashu) [reply controllers/managedocs_controller.go:176]: Tried this as well, didn't work

Comment (leelavg) [reply controllers/managedocs_controller.go:176]: @akshayashu 
- generally, for Read on api in CRUD (create, read, update, delete) ops you need `verbs=get;watch;list`
- rationale being, the sdk during setting up shared caches lists the resources and then performs ops on cache
- this should be the reason why you are seeing `forbidden` error mentioned in other comment

Comment (akshayashu) [reply controllers/managedocs_controller.go:176]: @leelavg Adding list didn't work too.

**Comment by akshayashu** [comment]
Resolved all the issues @nb-ohad @bindrad @leelavg @ezio-auditore

**Comment by leelavg** [comment]
- pls rebase, update and resolve conversation threads

File: controllers/managedocs_controller.go (hunk 6 above)
Comment (akshayashu) [reply controllers/managedocs_controller.go]: It worked with `verbs=get;watch;list`

File: controllers/managedocs_controller.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (akshayashu) [reply controllers/managedocs_controller.go]: resolved

File: controllers/managedocs_controller.go (hunk 6 above)
Comment (bindrad) [reply controllers/managedocs_controller.go]: @akshayashu Can you resolve this conversation?

**Review by bindrad — APPROVED** [review]

**Pull request merged**